go-controller/pkg/config/config.go

```diff
@@ -1735,10 +1768,31 @@ func completeGatewayConfig(allSubnets *configSubnets) error {
 	if err != nil || !utilnet.IsIPv6(v6IP) {
 		return fmt.Errorf("invalid gateway v6 join subnet specified, subnet: %s: error: %v", Gateway.V6JoinSubnet, err)
 	}
-
 	allSubnets.append(configSubnetJoin, v4JoinCIDR)
 	allSubnets.append(configSubnetJoin, v6JoinCIDR)
 
+	//validate v4 and v6 masquerade subnets
+	v4MasqueradeIP, v4MasqueradeCIDR, err := net.ParseCIDR(Gateway.V4MasqueradeSubnet)
+	if err != nil || utilnet.IsIPv6(v4MasqueradeIP) {
+		return fmt.Errorf("invalid gateway v4 masquerade subnet specified, subnet: %s: error: %v", Gateway.V4MasqueradeSubnet, err)
+	} else {
+		if err = allocateV4MasqueradeIPS(v4MasqueradeIP, masqueradeIPs); err != nil {
```

Comment:
So a few lines above we're separating the IP address (let's call it x) found in `Gateway.V4MasqueradeSubnet` from its prefix length and we're passing this IP address to `allocateV4MasqueradeIPS` , so that we can allocate x+1, x+2...x+5. This will work in most cases, that is when the input subnet was specified as `{network address}/{prefix length}`  but it won't work if the input subnet was given as `{any address in the subnet}/{prefix length}`. To account for this (unfortunate) case as well, I would use the IP returned in `v4MasqueradeCIDR`.

Reply:
Good catch here, thanks for pointing it out